File: sylow_solver/proof_tree.py
"""Tree-structured proof output for cleaner visualization."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Dict, List, Optional, Set, Tuple

if TYPE_CHECKING:
    from .facts import Disjunction, Fact


@dataclass
class DisjMeta:
    """Metadata for a disjunction to display it nicely."""

    var_name: str
    pred_name: str
    branch_values: List[str]


@dataclass
class FactInfo:
    """Simplified fact info for display."""

    label: str
    fact: "Fact"
    theorem: Optional[str]
    deps: List[str]


@dataclass
class CaseBranch:
    """A branch in a case split."""

    disj_id: str
    branch_idx: int
    label: str
    children: "ProofTree"


@dataclass
class ProofTree:
    """Tree structure for proof."""

    kind: str  # "empty", "fact", "case_split", "contradiction"
    fact_info: Optional[FactInfo] = None
    next_tree: Optional["ProofTree"] = None
    disj_id: Optional[str] = None
    var_name: Optional[str] = None
    branches: List[CaseBranch] = field(default_factory=list)
    explanation: Optional[str] = None


def infer_disj_meta(disj: "Disjunction") -> DisjMeta:
    """Infer human-readable metadata from a disjunction's facts."""
    facts = disj.facts
    if not facts:
        return DisjMeta("?", "false", [])

    pred_name = facts[0].name

    all_args = [f.args for f in facts]
    if not all_args or not all_args[0]:
        return DisjMeta("?", pred_name, [str(i) for i in range(len(facts))])

    transposed = list(zip(*all_args))
    varying_idx = None
    for i, col in enumerate(transposed):
        if len({str(x) for x in col}) > 1:
            varying_idx = i
            break

    var_name = _make_var_name(pred_name, varying_idx, facts[0].args)
    if varying_idx is not None:
        branch_values = [str(f.args[varying_idx]) for f in facts]
    else:
        branch_values = [str(i) for i in range(len(facts))]

    return DisjMeta(var_name, pred_name, branch_values)


def _make_var_name(pred_name: str, varying_idx: Optional[int], args: List) -> str:
    """Create a nice variable name for the varying argument."""
    subscript_map = str.maketrans("0123456789", "₀₁₂₃₄₅₆₇₈₉")

    if pred_name == "num_sylow" and varying_idx == 2:
        if len(args) >= 2:
            p = str(args[0])
            return f"n{p.translate(subscript_map)}"
        return "n"
    elif pred_name == "sylow_order" and varying_idx == 2:
        if len(args) >= 2:
            p = str(args[0])
            return f"ord{p.translate(subscript_map)}"
        return "ord"
    elif varying_idx is not None:
        return f"x{str(varying_idx).translate(subscript_map)}"
    else:
        return "case"


def build_proof_tree(
    facts: List["Fact"],
    disjunctions: List["Disjunction"],
    disj_meta: Dict[str, DisjMeta],
) -> ProofTree:
    """Build a proof tree from the environment facts."""
    grouped: Dict[frozenset[Tuple[str, int]], List["Fact"]] = {}
    for fact in facts:
        key = frozenset(fact.dis_ancestors)
        grouped.setdefault(key, []).append(fact)
    return _build_tree(disj_meta, frozenset(), grouped)


def _build_tree(
    meta_map: Dict[str, DisjMeta],
    current_case: frozenset[Tuple[str, int]],
    grouped_facts: Dict[frozenset[Tuple[str, int]], List["Fact"]],
) -> ProofTree:
    """Build tree recursively, grouping by case context."""
    matching = grouped_facts.get(current_case, [])
    if not matching and len(grouped_facts) == 0:
        return ProofTree(kind="empty")

    fact_infos = [
        FactInfo(
            label=fe.label or "?",
            fact=fe,
            theorem=fe.conc_thm.name if fe.conc_thm else None,
            deps=fe.dependencies,
        )
        for fe in matching
    ]

    has_contradiction = any(fi.fact.name == "false" for fi in fact_infos)

    if has_contradiction:
        if not fact_infos:
            return ProofTree(kind="empty")
        result = ProofTree(kind="contradiction", explanation="contradiction")
        for fi in reversed(fact_infos[:-1]):
            result = ProofTree(kind="fact", fact_info=fi, next_tree=result)
        return result

    next_split = _find_next_split(current_case, grouped_facts)
    if next_split is None:
        result = ProofTree(kind="empty")
        for fi in reversed(fact_infos):
            result = ProofTree(kind="fact", fact_info=fi, next_tree=result)
        return result

    disj_id, _branches = next_split
    meta = meta_map.get(disj_id, DisjMeta("case", "unknown", []))

    case_branches = []
    for idx, label in enumerate(meta.branch_values):
        new_case = frozenset(tuple(current_case | {(disj_id, idx)}))
        subtree = _build_tree(meta_map, new_case, grouped_facts)
        case_branches.append(CaseBranch(disj_id, idx, label, subtree))

    result = ProofTree(
        kind="case_split",
        disj_id=disj_id,
        var_name=meta.var_name,
        branches=case_branches,
    )
    for fi in reversed(fact_infos):
        result = ProofTree(kind="fact", fact_info=fi, next_tree=result)
    return result


def _find_next_split(
    current_case: frozenset[Tuple[str, int]],
    grouped_facts: Dict[frozenset[Tuple[str, int]], List["Fact"]],
) -> Optional[Tuple[str, List[int]]]:
    """Find the next case split point."""
    extensions: List[Tuple[str, int]] = []
    for anc in grouped_facts.keys():
        if current_case.issubset(anc):
            for d, i in anc - current_case:
                extensions.append((d, i))

    if not extensions:
        return None

    d, _ = extensions[0]
    branches = list(set(i for d2, i in extensions if d2 == d))
    return (d, branches)


def render_tree(
    facts: List["Fact"],
    disjunctions: List["Disjunction"],
    disj_meta: Dict[str, DisjMeta],
) -> List[str]:
    """Render the proof as ASCII art tree."""
    # Filter to essential facts
    essential = _get_essential_facts(facts)
    tree = build_proof_tree(essential, disjunctions, disj_meta)

    # Find contradictions for summary
    contradictions = [f for f in facts if f.name == "false"]
    summary = _render_summary(disj_meta, contradictions)

    lines = ["", "Proof Structure:", "════════════════", ""]
    lines.extend(_render_node(disj_meta, "", tree))
    lines.extend(summary)
    return lines


def _get_essential_facts(facts: List["Fact"]) -> List["Fact"]:
    """Get only facts that contribute to contradictions."""
    # Build label -> fact map
    fact_map = {f.label: f for f in facts if f.label}

    # Find all false() facts
    contradictions = [f for f in facts if f.name == "false"]

    # Trace back to find essential facts
    essential_labels: Set[str] = set()
    for contra in contradictions:
        _trace_essential(fact_map, contra.label, essential_labels)

    # Filter and deduplicate
    seen: Set[Tuple[Tuple, frozenset]] = set()
    result = []
    for f in facts:
        if f.label not in essential_labels:
            continue
        key = (f.name, tuple(f.args)), frozenset(f.dis_ancestors)
        if key not in seen:
            seen.add(key)
            result.append(f)
    return result


def _trace_essential(fact_map: Dict[str, "Fact"], label: Optional[str], visited: Set[str]) -> None:
    """Recursively trace dependencies to find essential facts."""
    if not label or label in visited:
        return
    visited.add(label)

    fact = fact_map.get(label)
    if fact:
        for dep in fact.dependencies:
            _trace_essential(fact_map, dep, visited)


def _render_node(meta: Dict[str, DisjMeta], prefix: str, tree: ProofTree) -> List[str]:
    """Render a single node of the proof tree."""
    if tree.kind == "empty":
        return []

    if tree.kind == "contradiction":
        return [f"{prefix}└─ ⊥ (contradiction)"]

    if tree.kind == "fact":
        fi = tree.fact_info
        assert fi is not None
        fact_str = f"{fi.fact.name}({', '.join(str(a) for a in fi.fact.args)})"
        thm_str = f"[{fi.theorem}]" if fi.theorem else "[hypothesis]"
        line = f"{prefix}• {fi.label}: {fact_str} {thm_str}"
        rest = _render_node(meta, prefix, tree.next_tree) if tree.next_tree else []
        return [line] + rest

    if tree.kind == "case_split":
        header = f"{prefix}┌─ Case split on {tree.var_name}:"
        lines = [header]
        total = len(tree.branches)
        for idx, branch in enumerate(tree.branches):
            is_last = idx == total - 1
            connector = "└" if is_last else "├"
            child_prefix = prefix + ("  " if is_last else "│ ")
            branch_header = f"{prefix}{connector}─ {branch.label}:"
            lines.append(branch_header)
            lines.extend(_render_node(meta, child_prefix, branch.children))
        return lines

    return []


def render_clean(
    facts: List["Fact"],
    disjunctions: List["Disjunction"],
    disj_meta: Dict[str, DisjMeta],
) -> List[str]:
    """Render cleaner linear output with improvements:
    1. Filter to essential facts only
    2. Group by case context with headers
    3. Add proof summary
    4. Highlight contradictions
    """
    # Get essential facts
    essential = _get_essential_facts(facts)

    # Group by case context
    grouped = _group_by_case_context(essential)

    # Render each group
    lines: List[str] = []
    for ctx, group_facts in grouped:
        lines.extend(_render_case_group(ctx, group_facts, disj_meta))

    # Add summary
    contradictions = [f for f in facts if f.name == "false"]
    lines.extend(_render_summary(disj_meta, contradictions))

    return lines


def _group_by_case_context(facts: List["Fact"]) -> List[Tuple[frozenset, List["Fact"]]]:
    """Group facts by their case context, sorted by context size."""
    contexts: Set[frozenset] = set()
    for f in facts:
        contexts.add(frozenset(f.dis_ancestors))

    sorted_contexts = sorted(contexts, key=len)

    return [
        (ctx, [f for f in facts if frozenset(f.dis_ancestors) == ctx]) for ctx in sorted_contexts
    ]


def _render_case_group(
    ctx: frozenset, facts: List["Fact"], disj_meta: Dict[str, DisjMeta]
) -> List[str]:
    """Render a case group with header."""
    if not ctx:
        header = ["═══ Hypotheses ═══", ""]
    else:
        ctx_str = _render_case_context(ctx, disj_meta)
        header = [f"═══ Case: {ctx_str} ═══", ""]

    body = []
    for f in facts:
        body.extend(_render_fact_clean(f, disj_meta))

    return header + body


def _render_case_context(ctx: frozenset, disj_meta: Dict[str, DisjMeta]) -> str:
    """Render a case context as a string."""
    parts = []
    for d, idx in sorted(ctx):
        meta = disj_meta.get(d)
        if meta and idx < len(meta.branch_values):
            parts.append(f"{meta.var_name}={meta.branch_values[idx]}")
        else:
            parts.append(f"{d}.{idx}")
    return ", ".join(parts)


def _render_fact_clean(fact: "Fact", disj_meta: Dict[str, DisjMeta]) -> List[str]:
    """Render a single fact with clean formatting."""
    is_false = fact.name == "false"
    rendered_args = ", ".join(str(a) for a in fact.args)
    fact_str = "⊥ (contradiction)" if is_false else f"{fact.name}({rendered_args})"
    label_str = fact.label or "?"

    if fact.conc_thm:
        thm_str = f"by {_pretty_theorem_name(fact.conc_thm.name)}"
    else:
        thm_str = "hypothesis"
    dep_str = ""
    if fact.dependencies:
        dep_str = " from " + " ".join(fact.dependencies)

    prefix = "  ✓ " if is_false else "  "
    return [
        f"{prefix}{label_str} : {fact_str}",
        f"{prefix}    {thm_str}{dep_str}",
        "",
    ]


def _render_summary(disj_meta: Dict[str, DisjMeta], contradictions: List["Fact"]) -> List[str]:
    """Render proof summary showing closed branches."""
    if not contradictions:
        return ["", "No contradictions found."]

    # Group by case context
    by_context: Dict[frozenset, List["Fact"]] = {}
    for c in contradictions:
        ctx = frozenset(c.dis_ancestors)
        by_context.setdefault(ctx, []).append(c)

    num_branches = len(by_context)
    plural = "" if num_branches == 1 else "es"

    lines = [
        "",
        "═══ Proof Summary ═══",
        "",
        f"Proved contradiction in {num_branches} branch{plural}:",
        "",
    ]

    for ctx in sorted(by_context.keys(), key=len):
        contra = by_context[ctx][0]
        ctx_str = _render_case_context(ctx, disj_meta) if ctx else "base case"
        thm_str = _pretty_theorem_name(contra.conc_thm.name) if contra.conc_thm else "hypothesis"
        lines.append(f"  ✓ {ctx_str} → {thm_str}")

    return lines


def _pretty_theorem_name(name: str) -> str:
    """Pretty-print theorem names."""
    name_map = {
        "sylow": "Sylow's theorem",
        "single_sylow_normal": "unique Sylow subgroup is normal",
        "not_simple": "simple ∧ not_simple → ⊥",
        "counting_contradiction": "element counting",
        "alternating_order": "order of alternating group",
        "alternating_simple": "Aₙ is simple (n≥5)",
        "lagrange": "Lagrange's theorem",
        "divides_contradiction": "divisibility contradiction",
        "embed_An": "embedding into Aₙ",
        "coset_action": "coset action",
        "simple_group_action": "simple group action",
        "count_order_pk_elements": "counting p^k-order elements",
        "subgroup_index": "subgroup index",
    }
    return name_map.get(name, name)

File: src/ProofTree.hs
-- | Tree-structured proof output for cleaner visualization.
--
-- This module builds a hierarchical tree from the proof trace, grouping facts
-- by their case context (disjunction branches). It provides both a cleaner
-- linear output (Option B) and a tree visualization (Option D).
module ProofTree
  ( ProofTree(..)
  , CaseBranch(..)
  , buildProofTree
  , renderTree
  , renderClean
  ) where

import Core (Fact(..), Label(..), DisjId(..), FactId(..), TheoremName, PredName(..)
            , theoremNameText, factName)
import Environment.Symbols (SymbolTable, symbolTable, ppFactWithSymbols)
import Environment.Types (ProofEnvironment, FactEntry(..), DisjMeta(..)
                         , feDisAncestors, feConcThm, feDependencies, peDisjMeta, peFacts)
import Data.List (intercalate, sortOn)
import qualified Data.IntMap.Strict as IntMap
import qualified Data.Set as Set

-- | A branch in a case split
data CaseBranch = CaseBranch
  { cbDisjId :: DisjId
  , cbBranchIdx :: Int
  , cbLabel :: String      -- e.g., "n₂ = 1"
  , cbChildren :: ProofTree
  } deriving (Show)

-- | Tree structure for proof
data ProofTree
  = PTEmpty
  | PTFact FactInfo ProofTree          -- A single fact followed by more
  | PTCaseSplit DisjId String [CaseBranch]  -- Disjunction with var name and branches
  | PTContradiction String             -- Reached false() with explanation
  deriving (Show)

-- | Simplified fact info for display
data FactInfo = FactInfo
  { fiLabel :: FactId
  , fiFact :: Fact
  , fiTheorem :: Maybe TheoremName
  , fiDeps :: [Label]
  } deriving (Show)

-- | Build a proof tree from the environment (filtered to essential facts)
buildProofTree :: ProofEnvironment -> ProofTree
buildProofTree env =
  let facts = reverse (peFacts env)  -- Chronological order
      metaMap = peDisjMeta env
      symTbl = symbolTable env

      -- Build lookup map and find essential facts
      factMap = IntMap.fromList [(unFactId (feLabel fe), fe) | fe <- facts]
      contradictions = [fe | fe <- facts, factName (feFact fe) == PFalse]
      essentialIds = Set.unions [traceEssential factMap (feLabel fe) | fe <- contradictions]
      essential = deduplicateFacts [fe | fe <- facts, Set.member (feLabel fe) essentialIds]

   in buildTree metaMap symTbl Set.empty essential

-- Build tree recursively, grouping by case context
buildTree :: IntMap.IntMap DisjMeta -> SymbolTable -> Set.Set (DisjId, Int) -> [FactEntry] -> ProofTree
buildTree _ _ _ [] = PTEmpty
buildTree metaMap tbl currentCase allFacts =
  let (matching, rest) = extractMatchingFacts currentCase allFacts
      factInfos = map toFactInfo matching
   in if hasContradiction factInfos
        then buildContradictionBranch factInfos
        else buildNormalBranch metaMap tbl currentCase factInfos rest

-- | Extract facts matching the current case context exactly
{-# INLINE extractMatchingFacts #-}
extractMatchingFacts :: Set.Set (DisjId, Int) -> [FactEntry] -> ([FactEntry], [FactEntry])
extractMatchingFacts currentCase = span (\fe -> feDisAncestors fe == currentCase)

-- | Convert a FactEntry to FactInfo for display
{-# INLINE toFactInfo #-}
toFactInfo :: FactEntry -> FactInfo
toFactInfo fe = FactInfo (feLabel fe) (feFact fe) (feConcThm fe) (feDependencies fe)

-- | Check if any fact is a contradiction (false)
{-# INLINE hasContradiction #-}
hasContradiction :: [FactInfo] -> Bool
hasContradiction = any (\fi -> factName (fiFact fi) == PFalse)

-- | Build a branch that ends in contradiction
buildContradictionBranch :: [FactInfo] -> ProofTree
buildContradictionBranch [] = PTEmpty
buildContradictionBranch fis = foldr PTFact (PTContradiction "contradiction") (init fis)

-- | Build a normal branch (possibly with case splits)
buildNormalBranch :: IntMap.IntMap DisjMeta -> SymbolTable
                  -> Set.Set (DisjId, Int) -> [FactInfo] -> [FactEntry] -> ProofTree
buildNormalBranch metaMap tbl currentCase factInfos rest =
  case findNextSplit currentCase rest of
    Nothing -> foldr PTFact PTEmpty factInfos
    Just (disjId, _) ->
      let caseBranches = buildCaseBranches metaMap tbl currentCase rest disjId
       in foldr PTFact (PTCaseSplit disjId (getVarName metaMap disjId) caseBranches) factInfos

-- | Get variable name from disjunction metadata
getVarName :: IntMap.IntMap DisjMeta -> DisjId -> String
getVarName metaMap disjId =
  case IntMap.lookup (unDisjId disjId) metaMap of
    Just meta -> dmVarName meta
    Nothing -> "case"

-- | Get branch values from disjunction metadata
getBranchValues :: IntMap.IntMap DisjMeta -> DisjId -> [String]
getBranchValues metaMap disjId =
  case IntMap.lookup (unDisjId disjId) metaMap of
    Just meta -> dmBranchValues meta
    Nothing -> map show [(0::Int)..]

-- | Build all case branches for a disjunction
buildCaseBranches :: IntMap.IntMap DisjMeta -> SymbolTable
                  -> Set.Set (DisjId, Int) -> [FactEntry] -> DisjId -> [CaseBranch]
buildCaseBranches metaMap tbl currentCase rest disjId =
  [ CaseBranch disjId idx label subtree
  | (idx, label) <- zip [0..] (getBranchValues metaMap disjId)
  , let newCase = Set.insert (disjId, idx) currentCase
        branchFacts = filter (\fe -> Set.isSubsetOf newCase (feDisAncestors fe)) rest
        subtree = buildTree metaMap tbl newCase branchFacts
  ]

-- | Find the next case split point
findNextSplit :: Set.Set (DisjId, Int) -> [FactEntry] -> Maybe (DisjId, [Int])
findNextSplit currentCase facts =
  case findExtensions currentCase facts of
    [] -> Nothing
    ((d, _) : exts) ->
      let branches = Set.toList $ Set.fromList [i | (d', i) <- (d, 0) : exts, d' == d]
       in Just (d, branches)

-- | Find disjunction extensions beyond current case
findExtensions :: Set.Set (DisjId, Int) -> [FactEntry] -> [(DisjId, Int)]
findExtensions currentCase facts =
  [ (d, i)
  | fe <- facts
  , (d, i) <- Set.toList (Set.difference (feDisAncestors fe) currentCase)
  ]

-- | Render the tree as ASCII art
renderTree :: ProofEnvironment -> [String]
renderTree env =
  let tree = buildProofTree env
      metaMap = peDisjMeta env
      symTbl = symbolTable env
      facts = reverse (peFacts env)
      contradictions = [fe | fe <- facts, factName (feFact fe) == PFalse]
      summary = renderSummary metaMap contradictions
   in ["", "Proof Structure:", "════════════════", ""]
      ++ renderNode symTbl metaMap "" tree
      ++ summary

renderNode :: SymbolTable -> IntMap.IntMap DisjMeta -> String -> ProofTree -> [String]
renderNode _ _ _ PTEmpty = []
renderNode _tbl _meta prefix (PTContradiction _) = [prefix ++ "└─ ⊥ (contradiction)"]
renderNode tbl meta prefix (PTFact fi rest) =
  let factStr = ppFactWithSymbols tbl (fiFact fi)
      labelStr = "F" ++ show (unFactId (fiLabel fi))
      thmStr = case fiTheorem fi of
                 Just tn -> " [" ++ theoremNameText tn ++ "]"
                 Nothing -> " [hypothesis]"
   in (prefix ++ "• " ++ labelStr ++ ": " ++ factStr ++ thmStr) : renderNode tbl meta prefix rest
renderNode tbl meta prefix (PTCaseSplit _disjId varName branches) =
  let header = prefix ++ "┌─ Case split on " ++ varName ++ ":"
   in header : concatMap (renderBranch tbl meta prefix (length branches)) (zip [0..] branches)

renderBranch :: SymbolTable -> IntMap.IntMap DisjMeta -> String -> Int -> (Int, CaseBranch) -> [String]
renderBranch tbl meta prefix total (idx, branch) =
  let isLast = idx == total - 1
      connector = if isLast then "└" else "├"
      childPrefix = if isLast then prefix ++ "  " else prefix ++ "│ "
      branchHeader = prefix ++ connector ++ "─ " ++ cbLabel branch ++ ":"
   in branchHeader : renderNode tbl meta childPrefix (cbChildren branch)

-- | Render cleaner linear output with improvements:
--   1. Filter to essential facts only (trace back from false())
--   2. Group by case context with headers
--   3. Add proof summary showing closed branches
--   4. Highlight contradictions
renderClean :: ProofEnvironment -> [String]
renderClean env =
  let facts = reverse (peFacts env)  -- Chronological order
      metaMap = peDisjMeta env
      symTbl = symbolTable env

      -- Build lookup map from FactId -> FactEntry
      factMap = IntMap.fromList [(unFactId (feLabel fe), fe) | fe <- facts]

      -- Find all false() facts (contradictions)
      contradictions = [fe | fe <- facts, factName (feFact fe) == PFalse]

      -- Trace back to find essential facts
      essentialIds = Set.unions [traceEssential factMap (feLabel fe) | fe <- contradictions]
      essential = [fe | fe <- facts, Set.member (feLabel fe) essentialIds]

      -- Deduplicate
      deduped = deduplicateFacts essential

      -- Group by case context
      grouped = groupByCaseContext deduped

      -- Render each group
      renderedGroups = concatMap (renderCaseGroup symTbl metaMap) grouped

      -- Build summary
      summary = renderSummary metaMap contradictions

   in renderedGroups ++ summary

-- Trace back through dependencies to find all essential facts
traceEssential :: IntMap.IntMap FactEntry -> FactId -> Set.Set FactId
traceEssential factMap startId = go Set.empty [startId]
  where
    go visited [] = visited
    go visited (fid:rest)
      | Set.member fid visited = go visited rest
      | otherwise = case IntMap.lookup (unFactId fid) factMap of
          Nothing -> go visited rest
          Just fe ->
            let depIds = [did | LFact did <- feDependencies fe]
             in go (Set.insert fid visited) (depIds ++ rest)

-- Deduplicate facts by (fact content, case context) while preserving order
deduplicateFacts :: [FactEntry] -> [FactEntry]
deduplicateFacts = go Set.empty
  where
    go _ [] = []
    go seen (fe:fes) =
      let key = (feFact fe, feDisAncestors fe)
       in if Set.member key seen
            then go seen fes
            else fe : go (Set.insert key seen) fes

-- Group facts by their case context
groupByCaseContext :: [FactEntry] -> [(Set.Set (DisjId, Int), [FactEntry])]
groupByCaseContext facts =
  let -- Get all unique case contexts, sorted by size (hypotheses first)
      contexts = Set.toList $ Set.fromList [feDisAncestors fe | fe <- facts]
      sortedContexts = sortOn Set.size contexts
      -- Group facts by context
   in [(ctx, [fe | fe <- facts, feDisAncestors fe == ctx]) | ctx <- sortedContexts]

-- Render a case group with header
renderCaseGroup :: SymbolTable -> IntMap.IntMap DisjMeta -> (Set.Set (DisjId, Int), [FactEntry]) -> [String]
renderCaseGroup symTbl metaMap (ctx, facts) =
  let header = if Set.null ctx
                 then ["═══ Hypotheses ═══", ""]
                 else ["═══ Case: " ++ renderCaseContext metaMap ctx ++ " ═══", ""]
      body = concatMap (renderFactClean symTbl metaMap) facts
   in header ++ body

-- Render case context as a string
renderCaseContext :: IntMap.IntMap DisjMeta -> Set.Set (DisjId, Int) -> String
renderCaseContext metaMap ctx =
  intercalate ", " [renderCase metaMap (d, i) | (d, i) <- Set.toList ctx]

-- Render a single fact with clean formatting
renderFactClean :: SymbolTable -> IntMap.IntMap DisjMeta -> FactEntry -> [String]
renderFactClean tbl _metaMap fe =
  let isFalse = factName (feFact fe) == PFalse
      factStr = if isFalse
                  then "⊥ (contradiction)"
                  else ppFactWithSymbols tbl (feFact fe)
      labelStr = "F" ++ show (unFactId (feLabel fe))
      thmStr = case feConcThm fe of
                 Just tn -> "by " ++ prettyTheoremName tn
                 Nothing -> "hypothesis"
      deps = feDependencies fe
      depStr = if null deps
                 then ""
                 else " from " ++ unwords (map labelText deps)
      prefix = if isFalse then "  ✓ " else "  "
   in [prefix ++ labelStr ++ " : " ++ factStr
      ,prefix ++ "    " ++ thmStr ++ depStr
      ,""]

-- Render proof summary
renderSummary :: IntMap.IntMap DisjMeta -> [FactEntry] -> [String]
renderSummary metaMap contradictions =
  let -- Group contradictions by case context
      grouped = groupByContext contradictions
      numBranches = length grouped
   in if null contradictions
        then ["", "No contradictions found."]
        else ["", "═══ Proof Summary ═══", ""
             ,"Proved contradiction in " ++ show numBranches ++ " branch" ++ (if numBranches == 1 then "" else "es") ++ ":"
             ,""] ++
             [summarizeBranch metaMap ctx fe
             | (ctx, fes) <- grouped
             , fe <- take 1 fes]

-- | Group facts by their case context, returning list of (context, facts) pairs
groupByContext :: [FactEntry] -> [(Set.Set (DisjId, Int), [FactEntry])]
groupByContext = foldr insertByCtx []
  where
    insertByCtx fe [] = [(feDisAncestors fe, [fe])]
    insertByCtx fe ((ctx, fes):rest)
      | feDisAncestors fe == ctx = (ctx, fe:fes) : rest
      | otherwise = (ctx, fes) : insertByCtx fe rest

-- Summarize a single branch
summarizeBranch :: IntMap.IntMap DisjMeta -> Set.Set (DisjId, Int) -> FactEntry -> String
summarizeBranch metaMap ctx fe =
  let ctxStr = if Set.null ctx
                 then "base case"
                 else renderCaseContext metaMap ctx
      thmStr = case feConcThm fe of
                 Just tn -> prettyTheoremName tn
                 Nothing -> "hypothesis"
   in "  ✓ " ++ ctxStr ++ " → " ++ thmStr

-- Render a case context element nicely
renderCase :: IntMap.IntMap DisjMeta -> (DisjId, Int) -> String
renderCase metaMap (DisjId d, idx) =
  case IntMap.lookup d metaMap of
    Just meta ->
      let val = safeIndex (dmBranchValues meta) idx (show idx)
       in dmVarName meta ++ "=" ++ val
    Nothing -> "D" ++ show d ++ "." ++ show idx

-- Safe list indexing with default value
safeIndex :: [a] -> Int -> a -> a
safeIndex xs i def
  | i < 0 = def
  | otherwise = go xs i
  where
    go [] _ = def
    go (x:_) 0 = x
    go (_:rest) n = go rest (n - 1)

labelText :: Label -> String
labelText (LFact (FactId n)) = "F" ++ show n
labelText (LDisj (DisjId n)) = "D" ++ show n

-- Pretty-print theorem names
prettyTheoremName :: TheoremName -> String
prettyTheoremName tn = case theoremNameText tn of
  "sylow" -> "Sylow's theorem"
  "single_sylow_normal" -> "unique Sylow subgroup is normal"
  "not_simple" -> "simple ∧ not_simple → ⊥"
  "counting_contradiction" -> "element counting"
  "alternating_order" -> "order of alternating group"
  "alternating_simple" -> "Aₙ is simple (n≥5)"
  "lagrange" -> "Lagrange's theorem"
  "divides_contradiction" -> "divisibility contradiction"
  "embed_An" -> "embedding into Aₙ"
  "coset_action" -> "coset action"
  "simple_group_action" -> "simple group action"
  "count_order_pk_elements" -> "counting p^k-order elements"
  "subgroup_index" -> "subgroup index"
  other -> other
